Restrictions on class patterns (match statement) are too strict / (possible cause: too eager to alias types)

**Bug Report**

Considering the following piece of code:

```python
NestedLines = tuple

def print_lines(ls: NestedLines, indent: int = 0) -> None:
    for line in ls:
        match line:
            case NestedLines():
                print_lines(line, indent+4)
            case _:
                print(" "*indent + str(line))

print_lines((1, [2,3,4], (5, [6,7]), 8))
```

`mypy --strict` complains at the `case NestedLines()` saying "Class pattern class must not be a type alias with type parameters". However, given that a class pattern has to be a type object (and not a genericalias object), there's no way to resolve this that makes both the runtime and the typechecker happy.

For example, if you define 

```python
from typing import TypeAlias
NestedLines: TypeAlias = tuple["NestedLines | object"]
```
Running this will result in a  `TypeError: called match pattern must be a class`.

**Expected Behavior**

Program should both typecheck AND print the following when run:
```
1
[2, 3, 4]
    5
    [6, 7]
8
```

Not sure what's the best solution... pyright is happy about this example, so perhaps some checks need to be removed/relaxed. In the (more complicated real usecase) code where I came accross this, the `x = tuple` was not intended as an alias, but there's no way to tell mypy "this is just an assignment", which I think would cmake more explicit that its usage in the match statement is not related to aliases.

The [documentation](https://mypy.readthedocs.io/en/stable/kinds_of_types.html#type-aliases) says "Because the distinction between an unannotated variable and a type alias is implicit, ambiguous or incorrect type alias declarations default to defining a normal variable instead of a type alias."; I would say this case is ambiguous/incorrect as a typealias, so it should be considered a regular variable definition. But I even tried `NestedLines: type = tuple`, and it is still taken as an alias. 

A possible desired behaviour would be: if the value used in a match statement is an "ambiguous" typealias (not defined with a TypeAlias/TypeStatement), and has actually been set to a type object (not a generic alias like tuple[Any]), do not emit an error

**Actual Behavior**

```
$ python3.12 -m mypy --strict example.py 
example.py:1: error: Missing type parameters for generic type "tuple"  [type-arg]
example.py:6: error: Class pattern class must not be a type alias with type parameters  [misc]
Found 2 errors in 1 file (checked 1 source file)
```

If I change it to a typealias

```
$ python3.12 example.py 
Traceback (most recent call last):
  File "/home/dmoissetdees/training-materials/unix/labs/whale/template/example.py", line 12, in <module>
    print_lines((1, [2,3,4], (5, [6,7]), 8))
  File "/home/dmoissetdees/training-materials/unix/labs/whale/template/example.py", line 7, in print_lines
    case NestedLines():
         ^^^^^^^^^^^^^
TypeError: called match pattern must be a class
```

And mypy still complains (in this case I think correctly):

```
example.py:7: error: Class pattern class must not be a type alias with type parameters  [misc]
```

**Your Environment**

- Mypy version used: mypy 1.8.0 (compiled: yes)
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12.2

As an aditional examples where the check on match statements is overly strict:

```python
def match_type(value: object, t: type[object]) -> bool:
    match value:
        case t():
            return True
        case _:
            return False

print(match_type(3, int))
print(match_type("3", int))
```

Result:
```
$ python3.12 -m mypy --strict example2.py 
example2.py:3: error: Expected type in class pattern; found "Type[builtins.object]"  [misc]
```

That looks like something that would clearly be allowed (even the error message "expected a type, found a type instead", sounds confusing). pyright accepts this one too.

--- a/mypy/checkexpr.py
+++ b/mypy/checkexpr.py
@@ -4975,7 +4975,7 @@ class ExpressionChecker(ExpressionVisitor[Type], ExpressionCheckerSharedApi):
         # This code can be only called either from checking a type application, or from
         # checking a type alias (after the caller handles no_args aliases), so we know it
         # was initially an IndexExpr, and we allow empty tuple type arguments.
-        if not validate_instance(fake, self.chk.fail, empty_tuple_index=True):
+        if not validate_instance(fake, self.chk.fail, indexed=True):
             fix_instance(
                 fake, self.chk.fail, self.chk.note, disallow_any=False, options=self.chk.options
             )

--- a/mypy/semanal.py
+++ b/mypy/semanal.py
@@ -4035,8 +4035,8 @@ class SemanticAnalyzer(
                 variadic = True
             new_tvar_defs.append(td)
 
-        empty_tuple_index = typ.empty_tuple_index if isinstance(typ, UnboundType) else False
-        return analyzed, new_tvar_defs, depends_on, empty_tuple_index
+        indexed = bool(isinstance(typ, UnboundType) and (typ.args or typ.empty_tuple_index))
+        return analyzed, new_tvar_defs, depends_on, indexed
 
     def is_pep_613(self, s: AssignmentStmt) -> bool:
         if s.unanalyzed_type is not None and isinstance(s.unanalyzed_type, UnboundType):
@@ -4135,10 +4135,10 @@ class SemanticAnalyzer(
             res = NoneType()
             alias_tvars: list[TypeVarLikeType] = []
             depends_on: set[str] = set()
-            empty_tuple_index = False
+            indexed = False
         else:
             tag = self.track_incomplete_refs()
-            res, alias_tvars, depends_on, empty_tuple_index = self.analyze_alias(
+            res, alias_tvars, depends_on, indexed = self.analyze_alias(
                 lvalue.name,
                 rvalue,
                 allow_placeholder=True,
@@ -4178,12 +4178,11 @@ class SemanticAnalyzer(
         no_args = (
             isinstance(res, ProperType)
             and isinstance(res, Instance)
-            and not res.args
-            and not empty_tuple_index
             and not pep_695
+            and not indexed
         )
         if isinstance(res, ProperType) and isinstance(res, Instance):
-            if not validate_instance(res, self.fail, empty_tuple_index):
+            if not validate_instance(res, self.fail, indexed):
                 fix_instance(res, self.fail, self.note, disallow_any=False, options=self.options)
         # Aliases defined within functions can't be accessed outside
         # the function, since the symbol table will no longer
@@ -5668,7 +5667,7 @@ class SemanticAnalyzer(
                 return
 
             tag = self.track_incomplete_refs()
-            res, alias_tvars, depends_on, empty_tuple_index = self.analyze_alias(
+            res, alias_tvars, depends_on, indexed = self.analyze_alias(
                 s.name.name,
                 s.value.expr(),
                 allow_placeholder=True,

--- a/mypy/typeanal.py
+++ b/mypy/typeanal.py
@@ -2160,6 +2160,15 @@ def instantiate_type_alias(
         # Note that we keep the kind of Any for consistency.
         return set_any_tvars(node, [], ctx.line, ctx.column, options, special_form=True)
 
+    if (
+        no_args
+        and isinstance(node.target, ProperType)
+        and isinstance(node.target, Instance)
+        and node.target.type.fullname == "builtins.tuple"
+        and len(args)
+    ):
+        no_args = False
+
     max_tv_count = len(node.alias_tvars)
     act_len = len(args)
     if (
@@ -2480,13 +2489,14 @@ def make_optional_type(t: Type) -> Type:
         return UnionType([t, NoneType()], t.line, t.column)
 
 
-def validate_instance(t: Instance, fail: MsgCallback, empty_tuple_index: bool) -> bool:
+def validate_instance(t: Instance, fail: MsgCallback, indexed: bool) -> bool:
     """Check if this is a well-formed instance with respect to argument count/positions."""
     # TODO: combine logic with instantiate_type_alias().
     if any(unknown_unpack(a) for a in t.args):
         # This type is not ready to be validated, because of unknown total count.
         # TODO: is it OK to fill with TypeOfAny.from_error instead of special form?
         return False
+    empty_tuple_index = indexed and not t.args
     if t.type.has_type_var_tuple_type:
         min_tv_count = sum(
             not tv.has_default() and not isinstance(tv, TypeVarTupleType)

--- a/mypy/types.py
+++ b/mypy/types.py
@@ -1043,6 +1043,7 @@ class UnboundType(ProperType):
         self,
         name: str,
         args: Sequence[Type] | None = None,
+        *,
         line: int = -1,
         column: int = -1,
         optional: bool = False,

--- a/test-data/unit/check-generics.test
+++ b/test-data/unit/check-generics.test
@@ -1002,7 +1002,7 @@ y: U
 reveal_type(x)  # N: Revealed type is "builtins.int | None"
 reveal_type(y)  # N: Revealed type is "builtins.int"
 
-U[int]  # E: Type application targets a non-generic function or class
+U[int]  # E: Bad number of arguments for type alias, expected 0, given 1
 O[int]  # E: Bad number of arguments for type alias, expected 0, given 1 \
         # E: Type application is only supported for generic classes
 

--- a/test-data/unit/check-python310.test
+++ b/test-data/unit/check-python310.test
@@ -1402,6 +1402,42 @@ def print_test(m: object, typ: type[T]) -> T:
             reveal_type(m)  # N: Revealed type is "builtins.object"
     raise
 
+[case testMatchClassPatternTupleAlias]
+# flags: --strict-equality --warn-unreachable
+from typing import Tuple
+
+tuple_alias = tuple
+Tuple_alias = Tuple
+
+def main(m: object):
+    match m:
+        case tuple():
+            reveal_type(m)  # N: Revealed type is "builtins.tuple[Any, ...]"
+        case _:
+            reveal_type(m)  # N: Revealed type is "builtins.object"
+
+    match m:
+        case tuple_alias():
+            reveal_type(m)  # N: Revealed type is "builtins.tuple[Any, ...]"
+        case _:
+            reveal_type(m)  # N: Revealed type is "builtins.object"
+
+    match m:
+        # With real typeshed you'll get an error like this instead:
+        # Expected type in class pattern; found "typing._SpecialForm"
+        case Tuple():  # E: Expected type in class pattern; found "builtins.int"
+            reveal_type(m)  # E: Statement is unreachable
+        case _:
+            reveal_type(m)  # N: Revealed type is "builtins.object"
+
+    match m:
+        # This is a false negative, it will raise at runtime
+        case Tuple_alias():
+            reveal_type(m)  # N: Revealed type is "builtins.tuple[Any, ...]"
+        case _:
+            reveal_type(m)  # N: Revealed type is "builtins.object"
+[builtins fixtures/tuple.pyi]
+
 [case testMatchNonFinalMatchArgs]
 class A:
     __match_args__ = ("a", "b")

